onnxruntime/test/providers/vitisai/vitisai_basic_test.cc

```diff
@@ -0,0 +1,316 @@
+  //
+  elapsedMicroseconds.QuadPart *= 1000000;
+  elapsedMicroseconds.QuadPart /= frequency.QuadPart;
+
+  return elapsedMicroseconds.QuadPart;
+}
+
+struct uploadReadback {
+  float upload, readback;
+};
+
+typedef struct uploadReadback Struct;
+
+template <typename TElement>
+Struct VAIExecutionProviderTest(TElement* indices, int x, int y, int z) { 
+
+  ComPtr<ID3D12Device> d3d12_device; 
+
+  int size = x * y * z;
+  
+  ORT_THROW_IF_FAILED(D3D12CreateDevice(nullptr, D3D_FEATURE_LEVEL_11_0, IID_PPV_ARGS(d3d12_device.ReleaseAndGetAddressOf())));
+
+  ComPtr<ID3D12CommandQueue> dxQueue;
+  ComPtr<ID3D12CommandAllocator> allocator;
+  ComPtr<ID3D12GraphicsCommandList> cmdList;
+
+  D3D12_COMMAND_QUEUE_DESC queueDesc = {};
+  queueDesc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+  queueDesc.Flags = D3D12_COMMAND_QUEUE_FLAG_NONE;
+
+  // create command queue 
+  ORT_THROW_IF_FAILED(d3d12_device->CreateCommandQueue(
+      &queueDesc, IID_PPV_ARGS(&dxQueue)));
+
+  // create command allocator 
+  ORT_THROW_IF_FAILED(d3d12_device->CreateCommandAllocator(
+      D3D12_COMMAND_LIST_TYPE_DIRECT,
+      IID_PPV_ARGS(allocator.GetAddressOf())));
+
+  // create command list 
+  ORT_THROW_IF_FAILED(d3d12_device->CreateCommandList(
+      0,
+      D3D12_COMMAND_LIST_TYPE_DIRECT,
+      allocator.Get(),  // Associated command allocator
+      nullptr,          // Initial PipelineStateObject
+      IID_PPV_ARGS(cmdList.GetAddressOf())));
+
+  size_t byteSize = size * sizeof(TElement);
+
+  ComPtr<ID3D12Resource> bufferUploader = nullptr;
+  ComPtr<ID3D12Resource> bufferGPU = nullptr;
+
+  // calling CPU->GPU function and measuring time
+  LARGE_INTEGER start = getStartingTime();
+  // copy CPU buffer to GPU
+  bufferGPU = vaip::tensor_proto_new_d3d12_cpu_to_gpu(d3d12_device.Get(), bufferUploader, cmdList.Get(),
+                                                      static_cast<void*>(indices), byteSize);
+  int uploadElapsed = getElapsedTime(start);
+
+  // copy GPU buffer to CPU (readback)  
+  void* output = nullptr;
+
+  // calling GPU->CPU readback function and measuring time
+  start = getStartingTime();
+  // copy GPU buffer to CPU (readback)
+  output = vaip::tensor_proto_new_d3d12_gpu_to_cpu(bufferGPU, d3d12_device.Get(), cmdList.Get(),
+                                                   byteSize, dxQueue);
+  
+  int readbackElapsed = getElapsedTime(start);
+
+  //// Write output to file
+  // std::ofstream fw("result_dst.txt", std::ofstream::out);
+  //// check if file was successfully opened for writing
+  // if (fw.is_open()) {
+  //   // store array contents to text file
+  //   for (int i = 0; i < x; i++) {
+  //     for (int j = 0; j < y; j++) {
+  //       for (int k = 0; k < z; k++) {
+  //         fw << "A[" << i << "][" << j << "][" << k << "] " << static_cast<int*>(output)[i + x * (j + y * k)];
+  //         fw << "\n";
+  //       }
+  //      
+  //     }
+  //   }
+  //   fw.close();
+  // }
+  free(output);  
```

Comment:
Please check if the `output` is a valid pointer (`!= NULL`) before using it and releasing it.